### ISSUE TITLE
ValueHolders are consuming too much memory

### ISSUE THREAD

**Issue by patric-r** [eclipse-ee4j/eclipselink#1941]
Some of our applications have PUs with statically weaved entities with many lazy OneToMany relationships (to be precise: a Set, so EclipseLink uses IndirectSet).

When fetching those entities we experience a huge memory consumption for the ValueHolder classes.
This already happens if do not access those IndirectSet (so those don't get instantiated)

E.g. for a real-life application with 16 GB heap, the situation looks really bad when analyzing the heap and showing the top-memory-consuming classes:

![image](https://github.com/eclipse-ee4j/eclipselink/assets/10393382/f79ff7d6-5407-4623-b53a-d68adb614869)

As you can see, from the top 16 memory consuming classes, eclipselink classes and their child instances (like most of the HashMap and Object[] instances as the GC roots prove) are occupying most the heap (more than 70% / 10GB) not taking the actual entity instances (including their backup clones) into account.

In this case, roughly 1GB of entity data (including backup clone, so in reality 500MB) have a management overhead of 8,5 GB (!) eclipselink class instances or instances held by those.
This is way too much and hints to architectural problems within EclipseLink.

The most severe issue (at least in our use-cases) are the excessive use of ValueHolders, taking up 4GB (!) of 16 GB.
While I understand that these are necessary for the lazy-loading mechanism, I don't understand why every indirection needs 2 instances (one instance of UnitOfWorkQueryHolder and nother instance of QueryBasedValueHolder, partially containing the same fields/information.
Every object increases heap demand and gc pressure and dramatically slows down our application.

Questions:
1. Is there a way to either use another, more efficient indirection technique or to lower the number of instances involved with indirection?

2. Is there a way to "disable" the indirection for certain entities (similar to the proprietary fetch groups), completely avoiding the ValueHolders at all for all entities where those are never accessed at all?

3. If the answers of questions 1 and 2 are "no" - do you agree with me that we need a better design here, at least reducing the number of instances from 2 per 1 per lazy entity reference?

**Comment by Zuplyx** [comment]
Hello @rfelcman, @patric-r 

I've created an example project for this issue. It has a main entity (``MyTestEntity``) with 6 ``OneToMany`` references.

For demonstration I filled the database with 200k entries. The attached picture shows a heap dump after executing a simple query fetching all of these into memory:

![heapDump_200k](https://github.com/eclipse-ee4j/eclipselink/assets/13404745/f5a6458d-40f3-43af-8874-1910a65f1f28)

As you can see this results in **1.2M instances each of ``UnitOfWorkQueryValueHolder`` and ``QueryBasedValueHolder``** and additionally **2.4M instances of ``IndirectSet``**.
This is a huge overhead for only 200k entities (and the additional 200k backup clones).

I think there are two relatively simple ways to decrease this overhead:

1. **Combine ``UnitOfWorkQueryValueHolder`` and ``QueryBasedValueHolder``:** Since ``UnitOfWorkQueryValueHolder`` wraps an instance of ``QueryBasedValueHolder``, maybe they could be combined into a single class which covers both use cases. This would cut the amount of ValueHolders required per reference into half.
2. **Reuse one instance of ``IndirectSet`` for the original and the clone:** As far as I see the original object currently uses an ``IndirectSet`` with a ``QueryBasedValueHolder`` and the clone uses another one with a ``UnitOfWorkQueryValueHolder``, which wraps the original ``QueryBasedValueHolder``. Since both of these instances share the same ValueHolder, we should investigate whether it's possible to use only one ``IndirectSet`` for both the original and the backup.  
  



The attached file contains the example maven project. It's readme contains more information on how to use it.

[jpa-bug-1941-ValueHoldersAreConsumingTooMuchMemory.zip](https://github.com/eclipse-ee4j/eclipselink/files/12673858/jpa-bug-1941-ValueHoldersAreConsumingTooMuchMemory.zip)

**Comment by rfelcman** [comment]
Hello,

I checked and tested attached test case and I have to confirm described behavior.

About questions above:
Questions:
---  Is there a way to either use another, more efficient indirection technique or to lower the number of instances involved with indirection?
NO

--- Is there a way to "disable" the indirection for certain entities (similar to the proprietary fetch groups), completely avoiding the ValueHolders at all for all entities where those are never accessed at all (during runtime for certain entity instances) to have a workaround at least?
NO only by OneToMany attribute removal

    If the answers of questions 1 and 2 are "no" - do you agree with me that we need a better design here, at least reducing the number of instances from 2 per 1 per lazy entity reference / indirection?
YES, but from my point of view this is corner case to keep a hundreds of thousands entities in managed stated in the active transaction.
There is some really old discussion about ValueHolders
https://bugs.eclipse.org/bugs/show_bug.cgi?id=278650
but it's not directly related to the issue above.
There some possible ways how to customize indirection, but it's about EclipseLink API no JPA
https://wiki.eclipse.org/Introduction_to_Mappings_(ELUG)#Transparent_Indirect_Container_Indirection
https://wiki.eclipse.org/Configuring_a_Mapping_(ELUG)#Configuring_Proxy_Indirection
https://wiki.eclipse.org/Configuring_a_Mapping_(ELUG)#Configuring_IndirectContainer_Indirection

Sorry, but due our limited resources we can't provide any solution soon.

**Comment by Zuplyx** [comment]
Hello @rfelcman,

did you have time to consider the changes I proposed in my comment? If so, do you think it is possible to implement them or are there any significant roadblocks I've overlooked?

We are willing to provide a PR for these changes, if you consider it realistic for them to be eventually merged. But we do not want to spend a potentially significant amount of time for a PR that is never going to be merged.

**Comment by rfelcman** [comment]
> Hello @rfelcman,
> 
> did you have time to consider the changes I proposed in my comment? If so, do you think it is possible to implement them or are there any significant roadblocks I've overlooked?
> 
I tested it one day, but didn't have time to make any POC yet due other urgent tasks from my employer.
> We are willing to provide a PR for these changes, if you consider it realistic for them to be eventually merged. But we do not want to spend a potentially significant amount of time for a PR that is never going to be merged.
Yes why not? If it will not break any current test.

**Comment by Zuplyx** [comment]
I created a POC for my first suggestion, combining `UnitOfWorkQueryValueHolder` and `QueryBasedValueHolder`. 
This approach will require deeper changes to eclipselink than I anticipated, especially in the area of change detection. Unfortunately I currently don't have the time to implement these.

I committed my changes along with some documentation to the [master_issue-1941](https://github.com/Zuplyx/eclipselink/tree/master_issue-1941) branch in my fork.

**Comment by cocorossello** [comment]
@Zuplyx Hi, are you using this patch?

**Comment by Zuplyx** [comment]
@cocorossello No, in order to work this would require deeper changes to Eclipselink's change detection mechanism. Unfortunately I don't have the time to work on this further.
